File: src/main/java/com/thunderbear06/config/ConfigLoader.java
package com.thunderbear06.config;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.stream.JsonReader;
import net.neoforged.fml.loading.FMLPaths;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;

public final class ConfigLoader {
    private ConfigLoader() {
    }

    public static <T extends ConfigState> T loadConfig(String modID, T state) {
        Path configPath = FMLPaths.CONFIGDIR.get().resolve(modID + ".json");
        try {
            if (!Files.exists(configPath)) {
                Files.createDirectories(configPath.getParent());
                Files.writeString(configPath, new GsonBuilder().setPrettyPrinting().create().toJson(state));
                return state;
            }
            try (JsonReader reader = new JsonReader(Files.newBufferedReader(configPath))) {
                T loaded = new Gson().fromJson(reader, state.getClass());
                return loaded == null ? state : loaded;
            }
        } catch (IOException e) {
            throw new RuntimeException("Failed to load config file " + configPath, e);
        }
    }

    public interface ConfigState {
    }
}

File: src/main/kotlin/com/thunderbear06/config/ConfigLoader.kt
package com.thunderbear06.config

import com.google.gson.Gson
import com.google.gson.GsonBuilder
import com.google.gson.stream.JsonReader
import net.neoforged.fml.loading.FMLPaths
import java.io.IOException
import java.nio.file.Files

object ConfigLoader {
    @JvmStatic
    fun <T : ConfigState> loadConfig(modID: String, state: T): T {
        val configPath = FMLPaths.CONFIGDIR.get().resolve("$modID.json")
        try {
            if (!Files.exists(configPath)) {
                Files.createDirectories(configPath.parent)
                Files.writeString(configPath, GsonBuilder().setPrettyPrinting().create().toJson(state))
                return state
            }

            JsonReader(Files.newBufferedReader(configPath)).use { reader ->
                val loaded: T? = Gson().fromJson(reader, state.javaClass)
                return loaded ?: state
            }
        } catch (e: IOException) {
            throw RuntimeException("Failed to load config file $configPath", e)
        }
    }

    interface ConfigState
}
